File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0: return 0

        dp = [int(1e9) for i in range(amount + 1)]

        for i in range(1, amount + 1):
            for coin in coins:
                if coin == i: dp[i] = 1
                elif coin > i: continue
                else: dp[i] = min(dp[i], dp[i - coin] + 1)

        return dp[amount] if dp[amount] != int(1e9) else -1

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) {
            return 0;
        }

        int[] dp = new int[amount + 1];

        for (int i = 1; i < amount + 1; i++) {
            dp[i] = dp.length;

            for (int coin: coins) {
                if (coin == i) {
                    dp[i] = 1;
                } else if (coin < i) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        if (dp[amount] != dp.length) {
            return dp[amount];
        }

        return -1;
    }
}